vitrivr-engine-module-pgvector/src/main/kotlin/org/vitrivr/engine/database/pgvector/descriptor/PgDescriptorInitializer.kt

```diff
@@ -0,0 +1,142 @@
+        for (field in this.prototype.schema()) {
+            when (field.type) {
+                Type.String -> statement.append("\"${field.name}\" varchar(255), ")
+                Type.Text -> statement.append("\"${field.name}\" text, ")
+                Type.Boolean -> statement.append("\"${field.name}\" boolean, ")
+                Type.Byte -> statement.append("\"${field.name}\" smallint, ")
+                Type.Short -> statement.append("\"${field.name}\" smallint, ")
+                Type.Int -> statement.append("\"${field.name}\" integer, ")
+                Type.Long -> statement.append("\"${field.name}\" bigint, ")
+                Type.Float -> statement.append("\"${field.name}\" real, ")
+                Type.Double -> statement.append("\"${field.name}\" double precision, ")
+                Type.Datetime -> statement.append("\"${field.name}\" datetime, ")
+                is Type.BooleanVector -> statement.append("\"${field.name}\" bit(${field.type.dimensions}), ")
+                is Type.DoubleVector -> statement.append("\"${field.name}\" vector(${field.type.dimensions}), ")
+                is Type.FloatVector -> statement.append("\"${field.name}\" vector(${field.type.dimensions}), ")
+                is Type.IntVector -> statement.append("\"${field.name}\" vector(${field.type.dimensions}), ")
+                is Type.LongVector -> statement.append("\"${field.name}\" vector(${field.type.dimensions}), ")
+            }
+        }
+
+        /* Finalize statement*/
+        statement.append("PRIMARY KEY ($DESCRIPTOR_ID_COLUMN_NAME), ")
+        statement.append("FOREIGN KEY ($RETRIEVABLE_ID_COLUMN_NAME) REFERENCES $RETRIEVABLE_ENTITY_NAME($RETRIEVABLE_ID_COLUMN_NAME) ON DELETE CASCADE);")
+
+        /* Create entity. */
+        try {
+            this.connection.jdbc.prepareStatement(/* sql = postgres */ statement.toString()).use {
+                it.execute()
+            }
+        } catch (e: SQLException) {
+            LOGGER.error(e) { "Failed to initialize entity '$tableName' due to exception." }
+            throw e
+        }
+
+        /* Create indexes (optional). */
+        for (index in this.field.indexes) {
+            try {
+                val indexStatement = when (index.type) {
+                    IndexType.SCALAR -> "CREATE INDEX ON $tableName (${index.attributes.joinToString(",")});"
+                    IndexType.FULLTEXT -> "CREATE INDEX ON $tableName USING gin(${index.attributes.joinToString(",") { "to_tsvector('${index.parameters["language"] ?: "english"}', $it)" }});"
+                    IndexType.NNS ->  {
+                        require(index.attributes.size == 1) { "NNS index can only be created on a single attribute." }
+                        val distance = index.parameters["distance"]?.let { Distance.valueOf(it.uppercase()) } ?: Distance.EUCLIDEAN
+                        "CREATE INDEX ON $tableName USING hnsw(${index.attributes.first()} ${distance.toIndexName()});"
```

Comment:
Same here, we might want to be able to configure what kind of index is to be used

Reply:
We could use a parameter here to make the distinction between IVFPQ and HNWS (which are the only two options).

Reply:
Those are the only two options without using additional extensions. Using https://github.com/timescale/pgvectorscale, for example, adds two additional index structures. I would not be surprised if there were others in the future that are compatible and do not require any additional implementation effort other than loading the extension. So, I think it would make sense to support specifying this with the parameters and setting a reasonable default.

Reply:
I guess the point I'm trying to make is a different one: This cannot just be addressed by allowing the user to specify parameters.

1. There is no such thing as an _a priori reasonable default_ for many index structures. The reasonable default depends on the data.
2. Some indexes are incapable of initialization without having access to data (e.g., PQ). Therefore, the fundamental flow of **initialize schema with indexes > import data > search** does not work in this case. Indexes must be (re-)created between data import and searching.

This would require fundamental changes to how indexes can be managed in vitrivr-engine, which is not in scope for this PR. Personally, I would argue that this is a task best left to the database layer OR some DBA.